### PR TITLE
PackageDir abstraction

### PR DESCRIPTION
**Pull request by isoos**
A step towards having a separate verification interface.

### CODE REVIEW

File: pkg/pub_package_reader/lib/src/package_dir.dart [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /// Free up resources (if it was extracted to a temp directory).
+  Future close();
+}
+
+/// Uses archive.tar.gz to provide access to the package archive.
+///
+/// Caches list of files for faster access.
+class ArchiveTarGzPackageDir extends PackageDir {
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/package_dir.dart]: Can we make the type private and just expose a function that takes a `tarballPath` and returns `PackageDir`?

File: pkg/pub_package_reader/lib/src/package_dir.dart [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'tar_utils.dart';
+
+/// Helper methods to access package content.
+abstract class PackageDir {
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/package_dir.dart]: Can we call it `PackageFileSystem` or `PackageContents` or `PackageFolder` ?

Comment (jonasfj) [reply pkg/pub_package_reader/lib/src/package_dir.dart]: What exceptions can this throw, any? It's fine if it only throws `Error` then we don't have document it.

let's also document that the path format, we use slashes, probably without stating slash...

File: pkg/pub_package_reader/lib/src/package_dir.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  /// List all the file names in the directory.
+  Future<List<String>> listFileNames();
+
+  /// Whether the file on the given relative [path] exists.
+  Future<bool> exists(String path);
+
+  /// Read the content of the file on the given relative [path].
+  Stream<List<int>> read(String path);
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/package_dir.dart]: What happens if the `path` doesn't exist?

File: pkg/pub_package_reader/lib/src/package_dir.dart [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+import 'tar_utils.dart';
+
+/// Helper methods to access package content.
+abstract class PackageDir {
+  /// List all the file names in the directory.
+  Future<List<String>> listFileNames();
+
+  /// Whether the file on the given relative [path] exists.
+  Future<bool> exists(String path);
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/package_dir.dart]: What happens for folders? symlinks? etc?

File: pkg/pub_package_reader/lib/src/tar_utils.dart [hunk 5]
```diff
@@ -26,21 +26,28 @@ Future<List<String>> listTarball(String path) async {
...
+/// Opens the file [name] to read (as binary).
+Stream<List<int>> openTarballFile(String path, String name) async* {
+  final p = await Process.start(
+    'tar',
+    ['-O', '-xzf', path, name],
+  );
+  yield* p.stdout;
+  final code = await p.exitCode;
+  if (code != 0) {
+    throw Exception('Failed to read tarball contents (code=$code).');
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/tar_utils.dart]: Exceptions thrown should be documented

File: pkg/pub_package_reader/lib/src/package_dir.dart (hunk 3 above)
Comment (jonasfj) [reply pkg/pub_package_reader/lib/src/package_dir.dart]: Can we just make this `Future<List<int>>` for simplicity?

File: pkg/pub_package_reader/lib/src/package_dir.dart (hunk 4 above)
Comment (isoos) [reply pkg/pub_package_reader/lib/src/package_dir.dart]: I think that's a fair concern, but I'm not sure how should I add to this part. Should it return true only if the file (and parent directories) are not symlinked?

File: pkg/pub_package_reader/lib/src/package_dir.dart (hunk 3 above)
Comment (isoos) [reply pkg/pub_package_reader/lib/src/package_dir.dart]: Should it also have a `maxLength` property to limit the memory use. Not sure if cleverly compressed archives are still a thing...

File: pkg/pub_package_reader/lib/src/package_folder.dart [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+  /// Whether the file on the given relative [path] exists.
+  ///
+  /// Throws IOException if directory listing or file access fails.
+  Future<bool> exists(String path);
+
+  /// Read the content of the file on the given relative [path].
+  ///
+  /// Emits IOException if directory listing or file access fails.
+  Stream<List<int>> read(String path);
```
Comment (jonasfj) [inline pkg/pub_package_reader/lib/src/package_folder.dart]: May I suggest that we just return `Uint8List`? and if the file doesn't exist we return `null`? It's a lot simpler?

Comment (jonasfj) [reply pkg/pub_package_reader/lib/src/package_folder.dart]: Then none of the methods here can throw any exceptions, we just return `null` if the file doesn't exist..

Comment (jonasfj) [reply pkg/pub_package_reader/lib/src/package_folder.dart]: If we have I/O error, then we throw it as a fatal error and crash..

**Comment by isoos** [comment]
Closing for now, let's do a `package:file` version of the same.

**Pull request closed without merging**